src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/CmsSigner.cs

```diff
@@ -224,22 +224,30 @@ internal void CheckCertificateValue()
                 newSignerInfo.UnsignedAttributes = PkcsHelpers.NormalizeAttributeSet(attrs.ToArray());
             }
 
-            bool signed = CmsSignature.Sign(
-                dataHash,
-                hashAlgorithmName,
-                Certificate,
-                PrivateKey,
-                silent,
-                out Oid signatureAlgorithm,
-                out ReadOnlyMemory<byte> signatureValue);
-
-            if (!signed)
+            if (SignerIdentifierType != SubjectIdentifierType.NoSignature)
             {
-                throw new CryptographicException(SR.Cryptography_Cms_CannotDetermineSignatureAlgorithm);
-            }
+                bool signed = CmsSignature.Sign(
+                    dataHash,
+                    hashAlgorithmName,
+                    Certificate,
+                    PrivateKey,
+                    silent,
+                    out Oid signatureAlgorithm,
+                    out ReadOnlyMemory<byte> signatureValue);
+
+                if (!signed)
+                {
+                    throw new CryptographicException(SR.Cryptography_Cms_CannotDetermineSignatureAlgorithm);
+                }
 
-            newSignerInfo.SignatureValue = signatureValue;
-            newSignerInfo.SignatureAlgorithm.Algorithm = signatureAlgorithm;
+                newSignerInfo.SignatureValue = signatureValue;
+                newSignerInfo.SignatureAlgorithm.Algorithm = signatureAlgorithm;
+            }
+            else
+            {
+                newSignerInfo.SignatureValue = dataHash;
+                newSignerInfo.SignatureAlgorithm.Algorithm = new Oid(Oids.NoSignature, null);
```

Comment:
I have zero idea whether this is correct. There's no unit test for it either. That definitely needs to be addressed.

Reply:
I'd not add a support if we are not sure it is correct

Reply:
The point is that it's halfway implemented and broken. The correct thing is to add test and fix it, which I plan to do. Only then it makes sense to implement the exception checks and test them.

Reply:
As long as there are tests I'm fine with this, if not planning to add tests I think we can make it an exception